File: client/ts/src/framer/streamer.ts
// Copyright 2026 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

import {
  EOF,
  type Stream,
  Unreachable,
  type WebSocketClient,
} from "@synnaxlabs/freighter";
import { errors, Rate, sync, TimeSpan, zod } from "@synnaxlabs/x";
import { z } from "zod";

import { type channel } from "@/channel";
import { paramsZ } from "@/channel/payload";
import { type ChannelRetriever, ReadAdapter } from "@/framer/adapter";
import { WSStreamerCodec } from "@/framer/codec";
import { Frame, frameZ } from "@/framer/frame";
import { StreamProxy } from "@/framer/streamProxy";

const reqZ = z.object({
  keys: z.number().array(),
  downsampleFactor: z.int(),
  throttleRate: Rate.z.optional(),
  excludeGroups: z.uint32().array().optional(),
  keepalive: TimeSpan.z.optional(),
});

/**
 * Request interface for streaming frames from a Synnax cluster.
 * Contains the keys of channels to stream from and a downsample factor.
 */
export interface StreamerRequest extends z.infer<typeof reqZ> {}

const resZ = z.object({
  frame: frameZ,
  /** Marks an empty response the Core emits so a dead connection is detectable. */
  keepalive: z.boolean().optional(),
});

/**
 * Response interface for streaming frames from a Synnax cluster.
 * Contains a frame of telemetry data.
 */
export interface StreamerResponse extends z.infer<typeof resZ> {}

const intermediateStreamerConfigZ = z.object({
  /** The channels to stream data from. Can be channel keys, names, or payloads. */
  channels: paramsZ,
  /** Optional factor to downsample the data by. Defaults to 1 (no downsampling). */
  downsampleFactor: z.int().default(1),
  /** Optional throttle rate in Hz to limit the rate of frames sent to the client. Defaults to 0 (no throttling). */
  throttleRate: Rate.z.default(new Rate(0)),
  /** excludeGroups sets writer group IDs whose frames should be filtered out by the
   Core. Used for telemetry bypass deduplication. */
  excludeGroups: z.uint32().array().default([]),
  /** Interval at which the Core emits keepalive responses so a silently dead
   connection fails reads instead of hanging forever. TimeSpan.ZERO disables
   detection. Defaults to 5 seconds. */
  keepalive: TimeSpan.z.default(TimeSpan.seconds(5)),
});

/** Zod schema for {@link StreamerConfig}. A bare channel list parses as a config. */
export const streamerConfigZ = intermediateStreamerConfigZ.or(
  paramsZ.transform((channels) => intermediateStreamerConfigZ.parse({ channels })),
);

/** Config for a streamer. Pass it to `client.telem.openStreamer`. */
export type StreamerConfig = z.input<typeof streamerConfigZ>;

/**
 * Streams frames of telemetry from a Synnax cluster in real time. Open one with the
 * client's openStreamer method, never directly. Read frames with `read` or by
 * iterating the streamer. Close it in a `finally` block to free its resources.
 *
 * @see https://docs.synnaxlabs.com/reference/client/working-with-data/streaming-data
 */
export interface Streamer extends AsyncIterator<Frame>, AsyncIterable<Frame> {
  /** The keys of the channels currently being streamed from. */
  keys: channel.Key[];
  /** Replaces the list of channels being streamed from. */
  update: (channels: channel.Params) => Promise<void>;
  /** Close the streamer and free all associated resources. */
  close: () => void;
  /**
   * Read the next frame of telemetry.
   * @throws {Unreachable} if keepalives were flowing and the stream then stays silent
   * past the keepalive deadline: the connection is presumed dead.
   */
  read: () => Promise<Frame>;
}

/** A function that opens a streamer. */
export interface StreamOpener {
  (config: StreamerConfig): Promise<Streamer>;
}

// Deadline for the Core to acknowledge a streamer once the socket is open. The
// handshake already proved the connection, so only a death in the window between the
// two reaches this.
const OPEN_ACK_TIMEOUT = TimeSpan.seconds(30);

/**
 * Creates a function that opens streamers with the given channel resolver and client.
 * @param retrieveChannels - Resolves channel params to payloads for the codec
 * @param client - The WebSocket client to use for streaming
 * @returns A function that opens streamers with the given configuration
 */
export const createStreamOpener =
  (retrieveChannels: ChannelRetriever, client: WebSocketClient): StreamOpener =>
  async (config) => {
    const cfg = zod.parse(streamerConfigZ, config, { label: "streamer config" });
    const adapter = await ReadAdapter.open(retrieveChannels, cfg.channels);
    client = client.withCodec(new WSStreamerCodec(adapter.codec));
    const stream = await client.stream("/frame/stream", reqZ, resZ);
    const streamer = new BaseStreamer(
      stream,
      adapter,
      cfg.downsampleFactor,
      cfg.throttleRate,
      cfg.excludeGroups,
      cfg.keepalive,
    );
    stream.send({
      keys: Array.from(adapter.keys),
      downsampleFactor: cfg.downsampleFactor,
      throttleRate: cfg.throttleRate,
      excludeGroups: cfg.excludeGroups,
      keepalive: cfg.keepalive,
    });
    // A keepalive can beat the open ack onto the wire, so the ack is the first
    // non-keepalive response.
    const ack = (async () => {
      let res = await stream.receive();
      while (res.keepalive === true) res = await stream.receive();
    })();
    const span = OPEN_ACK_TIMEOUT.toString();
    try {
      await sync.withTimeout(
        ack,
        OPEN_ACK_TIMEOUT,
        () =>
          new Unreachable({
            message: `streamer was not acknowledged within ${span}`,
          }),
      );
    } catch (err) {
      streamer.close();
      throw errors.fromUnknown(err);
    }
    return streamer;
  };

/**
 * Opens a new streamer with the given configuration.
 * @param retrieveChannels - Resolves channel params to payloads for the codec
 * @param client - The WebSocket client to use for streaming
 * @returns A promise that resolves to a new streamer
 */
export const openStreamer = async (
  retrieveChannels: ChannelRetriever,
  client: WebSocketClient,
  config: StreamerConfig,
): Promise<Streamer> => await createStreamOpener(retrieveChannels, client)(config);

// Missing this many keepalive intervals in a row fails the pending read: one is normal
// jitter, three is a dead connection.
const KEEPALIVE_DEADLINE_FACTOR = 3;

class BaseStreamer implements Streamer {
  private readonly stream: StreamProxy<typeof reqZ, typeof resZ>;
  private readonly adapter: ReadAdapter;
  private readonly downsampleFactor: number;
  private readonly throttleRate: Rate;
  private readonly excludeGroups: number[];
  private readonly deadline: TimeSpan;
  // Set once the Core proves keepalive support by sending one, so the deadline never
  // arms against a Core that will not send them.
  private armed = false;

  constructor(
    stream: Stream<typeof reqZ, typeof resZ>,
    adapter: ReadAdapter,
    downsampleFactor: number = 1,
    throttleRate: Rate = new Rate(0),
    excludeGroups: number[] = [],
    keepalive: TimeSpan = TimeSpan.ZERO,
  ) {
    this.stream = new StreamProxy("Streamer", stream);
    this.adapter = adapter;
    this.downsampleFactor = downsampleFactor;
    this.throttleRate = throttleRate;
    this.excludeGroups = excludeGroups;
    this.deadline = TimeSpan.milliseconds(
      keepalive.milliseconds * KEEPALIVE_DEADLINE_FACTOR,
    );
  }

  get keys(): channel.Key[] {
    return Array.from(this.adapter.keys);
  }

  async next(): Promise<IteratorResult<Frame, any>> {
    try {
      const frame = await this.read();
      return { done: false, value: frame };
    } catch (err) {
      if (EOF.matches(err)) return { done: true, value: undefined };
      throw errors.fromUnknown(err);
    }
  }

  async read(): Promise<Frame> {
    while (true) {
      const res = await this.receiveWithDeadline();
      if (res.keepalive === true) {
        if (!this.deadline.isZero) this.armed = true;
        continue;
      }
      return this.adapter.adapt(new Frame(res.frame));
    }
  }

  private async receiveWithDeadline(): Promise<z.infer<typeof resZ>> {
    const received = this.stream.receive();
    if (!this.armed) return await received;
    let timer: ReturnType<typeof setTimeout> | undefined;
    const deadline = new Promise<never>((_, reject) => {
      timer = setTimeout(() => {
        const silence = this.deadline.toString();
        const message = `streamer received no response for ${silence}`;
        reject(new Unreachable({ message }));
      }, this.deadline.milliseconds);
    });
    try {
      return await Promise.race([received, deadline]);
    } catch (err) {
      // The read already failed for its caller; a late settle of the losing receive
      // must not surface as an unhandled rejection.
      received.catch(() => {});
      throw errors.fromUnknown(err);
    } finally {
      clearTimeout(timer);
    }
  }

  async update(channels: channel.Params): Promise<void> {
    const hasChanged = await this.adapter.update(channels);
    if (!hasChanged) return;
    this.stream.send({
      keys: Array.from(this.adapter.keys),
      downsampleFactor: this.downsampleFactor,
      throttleRate: this.throttleRate,
      excludeGroups: this.excludeGroups,
    });
  }

  close(): void {
    this.stream.closeSend();
  }

  [Symbol.asyncIterator](): AsyncIterator<Frame, any, undefined> {
    return this;
  }
}

File: synnax/pkg/service/framer/streamer/streamer.go
// Copyright 2025 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

package streamer

import (
	"context"

	"github.com/synnaxlabs/alamos"
	"github.com/synnaxlabs/synnax/pkg/distribution/channel"
	"github.com/synnaxlabs/synnax/pkg/distribution/framer"
	"github.com/synnaxlabs/synnax/pkg/service/framer/calculation"
	"github.com/synnaxlabs/x/address"
	"github.com/synnaxlabs/x/config"
	"github.com/synnaxlabs/x/confluence"
	"github.com/synnaxlabs/x/confluence/plumber"
	"github.com/synnaxlabs/x/override"
	"github.com/synnaxlabs/x/telem"
	"github.com/synnaxlabs/x/validate"
)

type (
	Request         = Config
	Response        = framer.StreamerResponse
	Streamer        = confluence.Segment[Request, Response]
	responseSegment = confluence.Segment[Response, Response]
)

type Config struct {
	Keys             channel.Keys `json:"keys" msgpack:"keys"`
	SendOpenAck      bool         `json:"send_open_ack" msgpack:"send_open_ack"`
	DownSampleFactor int          `json:"down_sample_factor" msgpack:"down_sample_factor"`
	ThrottleRate     telem.Rate   `json:"throttle_rate" msgpack:"throttle_rate"`
}

func (cfg Config) distribution() framer.StreamerConfig {
	return framer.StreamerConfig{
		Keys:        cfg.Keys,
		SendOpenAck: cfg.SendOpenAck,
	}
}

// ServiceConfig is the configuration for opening a new streamer service.
type ServiceConfig struct {
	alamos.Instrumentation
	Calculation *calculation.Service `json:"calculation" msgpack:"calculation"`
	Channel     channel.Readable     `json:"channel" msgpack:"channel"`
	Framer      *framer.Service      `json:"framer" msgpack:"framer"`
}

var (
	_                    config.Config[ServiceConfig] = ServiceConfig{}
	DefaultServiceConfig                              = ServiceConfig{}
)

func (cfg ServiceConfig) Override(other ServiceConfig) ServiceConfig {
	cfg.Instrumentation = override.Zero(cfg.Instrumentation, other.Instrumentation)
	cfg.Calculation = override.Nil(cfg.Calculation, other.Calculation)
	cfg.Channel = override.Nil(cfg.Channel, other.Channel)
	cfg.Framer = override.Nil(cfg.Framer, other.Framer)
	return cfg
}

func (cfg ServiceConfig) Validate() error {
	v := validate.New("streamer")
	validate.NotNil(v, "calculation", cfg.Calculation)
	validate.NotNil(v, "channel", cfg.Channel)
	validate.NotNil(v, "framer", cfg.Framer)
	return v.Error()
}

type Service struct {
	cfg ServiceConfig
}

func NewService(cfgs ...ServiceConfig) (*Service, error) {
	cfg, err := config.New(DefaultServiceConfig, cfgs...)
	return &Service{cfg: cfg}, err
}

var (
	distAddr       address.Address = "distribution"
	utAddr         address.Address = "updater-transform"
	downSampleAddr address.Address = "down-sample"
	throttleAddr   address.Address = "throttle"
)

func (s *Service) New(ctx context.Context, cfg Config) (Streamer, error) {
	p := plumber.New()
	dist, err := s.cfg.Framer.NewStreamer(ctx, cfg.distribution())
	if err != nil {
		return nil, err
	}
	plumber.SetSegment(p, distAddr, dist)
	ut, err := s.newUpdaterTransform(ctx, cfg)
	if err != nil {
		return nil, err
	}
	plumber.SetSegment(p, utAddr, ut)
	plumber.MustConnect[framer.StreamerRequest](p, utAddr, distAddr, 25)
	var routeOutletFrom = distAddr
	if cfg.ThrottleRate > 0 {
		plumber.SetSegment(p, throttleAddr, newThrottle(cfg))
		plumber.MustConnect[Response](p, routeOutletFrom, throttleAddr, 25)
		routeOutletFrom = throttleAddr
	}
	if cfg.DownSampleFactor > 0 {
		plumber.SetSegment(p, downSampleAddr, newDownSampler(cfg))
		plumber.MustConnect[Response](p, routeOutletFrom, downSampleAddr, 25)
		routeOutletFrom = downSampleAddr
	}
	return &plumber.Segment[Request, Response]{
		Pipeline:         p,
		RouteInletsTo:    []address.Address{utAddr},
		RouteOutletsFrom: []address.Address{routeOutletFrom},
	}, nil
}

func (s *Service) newUpdaterTransform(
	ctx context.Context,
	cfg Config,
) (confluence.Segment[Request, framer.StreamerRequest], error) {
	ut := &calculationUpdaterTransform{
		Instrumentation: s.cfg.Instrumentation,
		c:               s.cfg.Calculation,
		readable:        s.cfg.Channel,
	}
	ut.Transform = ut.transform
	return ut, ut.update(ctx, cfg.Keys)
}
